### PR TITLE
SparkplugB feature addition to agent 

### PR DESCRIPTION
**Pull request by yshashix**
SparkplugB functionality added with how to enable in README file

### CODE REVIEW

**Review by srware — CHANGES_REQUESTED** [review]

File: config/config.json.template [hunk 1]
```diff
@@ -32,7 +32,13 @@
...
 			"retain": false,
 			"secure": true,
-			"retries": 5
+			"retries": 5,
+			"strictSSL": false,
+			"sparkplugB": true, 
+            "version": "spBv1.0",
+            "hwVersion": "Emulated Hardware",
+            "swVersion": "v1.0.0",
+            "rejectUnauthorized": false
```
[the remark this answers is not in the document]
Comment (srware) [reply config/config.json.template:41]: Can we re-use the 'secure' or 'strictSSL' config options which already exist for MQTT?

[the remark this answers is not in the document]
Comment (yshashix) [reply config/config.json.template:41]: Yes, I checked and it's the same at the end.

File: config/config.json.template [hunk 2]
```diff
@@ -32,7 +32,13 @@
 			"qos": 1,
 			"retain": false,
 			"secure": true,
-			"retries": 5
+			"retries": 5,
+			"strictSSL": false,
+			"sparkplugB": true, 
+            "version": "spBv1.0",
+            "hwVersion": "Emulated Hardware",
+            "swVersion": "v1.0.0",
```
[the remark this answers is not in the document]
Comment (yshashix) [reply config/config.json.template:40]: Currently, i don't see commands to set attributes of the device at agent, in data/device.json though there is api for it.
So for now I set default value for this parameter and will update code once it is added in device.json or any other file.

File: lib/cloud.proxy.js [hunk 3]
```diff
@@ -53,13 +53,73 @@ function IoTKitCloud(logger, deviceId, customProxy) {
...
-        if (deviceConf.device_id && deviceConf.device_token) {
-            me.mqttProxy.updateDeviceInfo(deviceConf);
-        } else {
-            me.logger.info("No credentials found for MQTT!");
-            me.mqttProxy = undefined;
+    /** 
+    * Checking if SparkplugB is also enabled or not
+    */
+        if (conf.connector.mqtt.sparkplugB) {
+            me.spbMetricList= [];
```
Comment (scooter1556) [inline lib/cloud.proxy.js:60]:
```suggestion
            me.spbMetricList = [];
```

File: lib/cloud.proxy.js [hunk 4]
```diff
@@ -53,13 +53,73 @@ function IoTKitCloud(logger, deviceId, customProxy) {
...
+            deviceConf['sensor_list'].forEach(comp => {               
+                utils.getItemFromCatalog(comp.type, function (item) {               
+                    if (item) {
+                        var compMetric= {
```
Comment (scooter1556) [inline lib/cloud.proxy.js:64]:
```suggestion
                        var compMetric = {
```

File: lib/cloud.proxy.js [hunk 5]
```diff
@@ -53,13 +53,73 @@ function IoTKitCloud(logger, deviceId, customProxy) {
...
+                            "alias"  : comp.cid ,
+                            "timestamp" : new Date().getTime(),
+                            "dataType" : item.format,
+                            "value" : 0
+                        }
+                        me.spbMetricList.push(compMetric);                     
+                    }
+                });                            
+            });
+            me.devProf ={
```
Comment (scooter1556) [inline lib/cloud.proxy.js:75]:
```suggestion
            me.devProf = {
```

File: lib/cloud.proxy.js [hunk 6]
```diff
@@ -53,13 +53,73 @@ function IoTKitCloud(logger, deviceId, customProxy) {
...
+                groupId   : deviceConf['account_id'],
+                edgeNodeId : deviceConf['gateway_id'],
+                clientId   : deviceConf['device_name'],
+                deviceId   : deviceConf['device_id'],         
+                componentMetric : me.spbMetricList
+            };
+
+            me.spBProxy = require('@open-iot-service-platform/oisp-sdk-js')(conf).lib.proxies.getSpBConnector();
+            me.logger.info("SparkplugB Mqtt proxy found! Configuring  Sparkplug and MQTT for data sending.");
```
Comment (scooter1556) [inline lib/cloud.proxy.js:84]:
```suggestion
            me.logger.info("SparkplugB MQTT proxy found! Configuring Sparkplug and MQTT for data sending.");
```

File: lib/cloud.proxy.js [hunk 7]
```diff
@@ -53,13 +53,73 @@ function IoTKitCloud(logger, deviceId, customProxy) {
...
+            if (deviceConf.device_token && me.spBProxy!= undefined) {
```
Comment (scooter1556) [inline lib/cloud.proxy.js:85]:
```suggestion
            if (deviceConf.device_token && me.spBProxy != undefined) {         
```

File: lib/cloud.proxy.js [hunk 8]
```diff
@@ -53,13 +53,73 @@ function IoTKitCloud(logger, deviceId, customProxy) {
...
+                me.logger.info("No credentials found for MQTT");
+                me.spBProxy = undefined;
+            }       
+            if (me.spBProxy != undefined ) { 
+                /** 
+                *  Sending Birth message for SparkplugB (of node/agent and for device),
+                * As per standard Birth message is mandatory to send on start before sending Data
+                */
+                me.spBProxy.nodeBirth(me.devProf, function(err) {                  
+                    if(err === "fail") {
```
Comment (scooter1556) [inline lib/cloud.proxy.js:97]:
```suggestion
                    if (err === "fail") {
```

File: lib/cloud.proxy.js [hunk 9]
```diff
@@ -53,13 +53,73 @@ function IoTKitCloud(logger, deviceId, customProxy) {
...
+                        me.logger.error("SparkplugB MQTT NBIRTH Metric not send, probably connection error ");
+                        me.spBProxy = undefined;
+                    }else{
```
Comment (scooter1556) [inline lib/cloud.proxy.js:100]:
```suggestion
                    } else {
```

File: lib/cloud.proxy.js [hunk 10]
```diff
@@ -53,13 +53,73 @@ function IoTKitCloud(logger, deviceId, customProxy) {
...
+                        me.logger.info("SparkplugB MQTT NBIRTH Metric sent succesfully for eonid: %j ",me.gatewayId );
```
Comment (scooter1556) [inline lib/cloud.proxy.js:101]:
```suggestion
                        me.logger.info("SparkplugB MQTT NBIRTH Metric sent succesfully for eonid: %j ", me.gatewayId);
```

File: lib/cloud.proxy.js [hunk 11]
```diff
@@ -53,13 +53,73 @@ function IoTKitCloud(logger, deviceId, customProxy) {
...
+                    }                                        
+                }); 
+                me.logger.debug("SparkplugB MQTT DBIRTH Metric: %j", me.spbMetricList);
+                me.spBProxy.deviceBirth(me.devProf, function(err) {                  
+                    if(err === "fail") {
```
Comment (scooter1556) [inline lib/cloud.proxy.js:106]:
```suggestion
                    if (err === "fail") {
```

File: lib/cloud.proxy.js [hunk 12]
```diff
@@ -53,13 +53,73 @@ function IoTKitCloud(logger, deviceId, customProxy) {
...
+                        me.logger.error("SparkplugB MQTT DBIRTH Metric not send, probably connection error ");
+                        me.spBProxy = undefined;
+                    }else{
```
Comment (scooter1556) [inline lib/cloud.proxy.js:109]:
```suggestion
                    } else {
```

File: lib/cloud.proxy.js [hunk 13]
```diff
@@ -53,13 +53,73 @@ function IoTKitCloud(logger, deviceId, customProxy) {
...
+                        me.logger.error("SparkplugB MQTT NBIRTH Metric not send, probably connection error ");
+                        me.spBProxy = undefined;
+                    }else{
+                        me.logger.info("SparkplugB MQTT NBIRTH Metric sent succesfully for eonid: %j ",me.gatewayId );
+                    }                                        
+                }); 
+                me.logger.debug("SparkplugB MQTT DBIRTH Metric: %j", me.spbMetricList);
+                me.spBProxy.deviceBirth(me.devProf, function(err) {                  
+                    if(err === "fail") {
+                        me.logger.error("SparkplugB MQTT DBIRTH Metric not send, probably connection error ");
```
Comment (scooter1556) [inline lib/cloud.proxy.js:107]:
```suggestion
                        me.logger.error("SparkplugB MQTT DBIRTH Metric not sent!");
```

File: lib/cloud.proxy.js [hunk 14]
```diff
@@ -53,13 +53,73 @@ function IoTKitCloud(logger, deviceId, customProxy) {
...
+                        me.spBProxy = undefined;
+                    }else{
+                        me.logger.info("SparkplugB MQTT DBIRTH Metric sent succesfully for device: %j ",me.deviceId);
```
Comment (scooter1556) [inline lib/cloud.proxy.js:110]:
```suggestion
                        me.logger.info("SparkplugB MQTT DBIRTH Metric sent succesfully for device: %j ", me.deviceId);
```

File: lib/cloud.proxy.js [hunk 15]
```diff
@@ -53,13 +53,73 @@ function IoTKitCloud(logger, deviceId, customProxy) {
...
+                    }                          
+                });       
+            }
+        }else{
```
Comment (scooter1556) [inline lib/cloud.proxy.js:114]:
```suggestion
        } else {
```

File: lib/cloud.proxy.js [hunk 16]
```diff
@@ -87,7 +147,7 @@ IoTKitCloud.prototype.isActivated = function () {
 IoTKitCloud.prototype.checkDeviceToken = function (callback) {
     var me = this,
         toCall = callback;
-
+
```
Comment (scooter1556) [inline lib/cloud.proxy.js:150]:
```suggestion

```

File: lib/cloud.proxy.js [hunk 17]
```diff
@@ -306,15 +365,59 @@ IoTKitCloud.prototype.disconnect = function () {
...
-    var me = this;
+    var me = this; 
+    var handler = function() {    
+        // SparkplugB format data submission to sdk
+        if (me.spBProxy != undefined) {
+            var componentMetrics = [];          
+            if (Array.isArray(metric)) {
+                metric.forEach(item => {
+                    let compMetric = me.spbMetricList.find(element => element.alias === item.alias);
+                    if(compMetric != undefined) {
```
Comment (scooter1556) [inline lib/cloud.proxy.js:376]:
```suggestion
                    if (compMetric != undefined) {
```

File: lib/cloud.proxy.js [hunk 18]
```diff
@@ -306,15 +365,59 @@ IoTKitCloud.prototype.disconnect = function () {
...
+                        compMetric.timestamp = item.on || new Date().getTime();
+                        componentMetrics.push(compMetric);
+                    }    
+                })
+                // Check if we have any data left to submit
+                if (componentMetrics.length === 0) {
+                    me.logger.error(' SPB Data submit - no data to submit.');
+                    return callback(false);
+                }
+            }else{
```
Comment (scooter1556) [inline lib/cloud.proxy.js:387]:
```suggestion
            } else { 
```

File: lib/cloud.proxy.js [hunk 19]
```diff
@@ -306,15 +365,59 @@ IoTKitCloud.prototype.disconnect = function () {
...
+                    }    
+                })
+                // Check if we have any data left to submit
+                if (componentMetrics.length === 0) {
+                    me.logger.error(' SPB Data submit - no data to submit.');
+                    return callback(false);
+                }
+            }else{ 
+                let compMetric = me.spbMetricList.find(element => element.alias === metric.alias);    
+                if(compMetric != undefined) {
```
Comment (scooter1556) [inline lib/cloud.proxy.js:389]:
```suggestion
                if (compMetric != undefined) {
```

File: lib/cloud.proxy.js [hunk 20]
```diff
@@ -306,15 +365,59 @@ IoTKitCloud.prototype.disconnect = function () {
...
+                    compMetric.value = metric.v;
+                    compMetric.timestamp = metric.on || new Date().getTime();
+                    componentMetrics.push(compMetric); 
+                }else {
```
Comment (scooter1556) [inline lib/cloud.proxy.js:393]:
```suggestion
                } else {
```

File: lib/cloud.proxy.js [hunk 21]
```diff
@@ -306,15 +365,59 @@ IoTKitCloud.prototype.disconnect = function () {
...
+                    me.logger.error('Data submission - could not find alias in stored metric');
+                    return callback(false);
+                }
+            }
+            me.logger.debug("SparkplugB MQTT DDATA Metric: %j", componentMetrics);
+            me.logger.debug("SparkplugB MQTT device profile: %j", me.devProf);
+            me.spBProxy.data(me.devProf,componentMetrics, function(err, response) {
+                if (err) {
+                    if(err === "fail") {
```
Comment (scooter1556) [inline lib/cloud.proxy.js:402]:
```suggestion
                    if (err === "fail") { 
```

File: lib/cloud.proxy.js [hunk 22]
```diff
@@ -306,15 +365,59 @@ IoTKitCloud.prototype.disconnect = function () {
...
+                    return callback(false);
+                }
+            }
+            me.logger.debug("SparkplugB MQTT DDATA Metric: %j", componentMetrics);
+            me.logger.debug("SparkplugB MQTT device profile: %j", me.devProf);
+            me.spBProxy.data(me.devProf,componentMetrics, function(err, response) {
+                if (err) {
+                    if(err === "fail") { 
+                        me.logger.error("SparkplugB MQTT DDATA Metric not send ");
+                    }else{
```
Comment (scooter1556) [inline lib/cloud.proxy.js:404]:
```suggestion
                        me.logger.error("SparkplugB MQTT DDATA Metric not sent!");
                    } else {
```

File: lib/cloud.proxy.js [hunk 23]
```diff
@@ -306,15 +365,59 @@ IoTKitCloud.prototype.disconnect = function () {
...
+                        me.logger.info("SparkplugB MQTT DDATA Metric sent successfully with payload %j",componentMetrics );
+                    }     
+                    callback(err)
```
Comment (scooter1556) [inline lib/cloud.proxy.js:407]:
```suggestion
                        me.logger.info("SparkplugB MQTT DDATA Metric sent successfully with payload %j", componentMetrics);
                    }     
                    callback(err);
```

File: lib/cloud.proxy.js [hunk 24]
```diff
@@ -306,15 +365,59 @@ IoTKitCloud.prototype.disconnect = function () {
...
+                } else {
+                    if (response) {
+                        callback(null, response)
```
Comment (scooter1556) [inline lib/cloud.proxy.js:410]:
```suggestion
                        callback(null, response);
```

File: lib/cloud.proxy.js [hunk 25]
```diff
@@ -306,15 +365,59 @@ IoTKitCloud.prototype.disconnect = function () {
...
+                    }
+                }
+            });
 
-    var handler = function() {
-        metric.accountId = me.secret.accountId;
-        metric.did = me.deviceId;
-        metric.gatewayId = me.gatewayId;
-        metric.deviceToken = me.secret.deviceToken;
+        }else if (me.mqttProxy != undefined) {
```
Comment (scooter1556) [inline lib/cloud.proxy.js:415]:
```suggestion
        } else if (me.mqttProxy != undefined) {
```

File: lib/cloud.proxy.js [hunk 26]
```diff
@@ -345,8 +448,8 @@ IoTKitCloud.prototype.dataSubmit = function (metric, callback) {
                         callback(null, response)
                     }
                 }
-            });
-            // ...until here
+            });
```
Comment (scooter1556) [inline lib/cloud.proxy.js:451]:
```suggestion
            }); 
```

File: lib/data.submission.js [hunk 27]
```diff
@@ -52,43 +52,75 @@ var Data = function (connector, SensorStore, logT) {
...
-            var metric = new Metric();
-
-            // Handle multiple observations
-            if (Array.isArray(msg)) {
-                for (var i = msg.length - 1; i >= 0; i -= 1) {
-                    // Check component id
-                    var cidArr = me.store.byName(msg[i].n);
-                    if (cidArr) {
-                        msg[i].cid = cidArr.cid; // Add component id to observation
+            if(config.connector.mqtt.sparkplugB) {
```
Comment (scooter1556) [inline lib/data.submission.js:55]:
```suggestion
            if (config.connector.mqtt.sparkplugB) {             
```

File: lib/data.submission.js [hunk 28]
```diff
@@ -52,43 +52,75 @@ var Data = function (connector, SensorStore, logT) {
...
-                        msg.splice(i, 1);
+                        me.logger.error('Data submission - could not find time series with the name: %s.', msg.n);
+                        return callback(true);
+                    }
+                }     
+                me.connector.dataSubmit(msg, function(dat) {
+                    me.logger.info("Response received: %j", dat);
+                    return callback(dat);
+                });
+            }else {
```
Comment (scooter1556) [inline lib/data.submission.js:86]:
```suggestion
            } else {
```

File: test/lib_data.submissionTests.js [hunk 29]
```diff
@@ -40,6 +40,33 @@ describe(fileToTest, function() {
     var store = {
 
     };
```
Comment (scooter1556) [inline test/lib_data.submissionTests.js:42]:
```suggestion
    var store = {};
```

File: test/lib_data.submissionTests.js [hunk 30]
```diff
@@ -40,6 +40,33 @@ describe(fileToTest, function() {
...
+    var conf ={
```
Comment (scooter1556) [inline test/lib_data.submissionTests.js:43]:
```suggestion
    var conf = {
```

File: test/lib_cloudproxyTest.js [hunk 31]
```diff
@@ -0,0 +1,462 @@
...
+        console.log(arguments);
+    };
+    var myDate = new Date();
+    var newDate = myDate.setHours(myDate.getHours()+1);
+    
+    var secret = {
+        'deviceTokenExpire': newDate
+    };
+    var store = {};
+    var conf ={
```
Comment (scooter1556) [inline test/lib_cloudproxyTest.js:48]:
```suggestion
    var conf = {
```

**Review by scooter1556 — CHANGES_REQUESTED** [review]

File: README.md [hunk 32]
```diff
@@ -109,6 +109,62 @@ To start the oisp-agent service simply execute the start script:
 
 For examples of how to use the 'oisp-agent' please see the [Examples](https://github.com/Open-IoT-Service-Platform/oisp-iot-agent/tree/master/examples) provided.
 
+### Enabling SparkPlugB Standard 
+
+1. To send data on SparkplugB standard data format, you need to enable it on agent and add its compatible sdk(oisp-sdk-js) to it. 
+
+For this you need to make certain changed on the config.json file as below in mqtt connector section (added in config.json.template):
```
Comment (srware) [inline README.md:116]:
```suggestion
1. To enable SparkplugB standard data format, you need to update the agent config as below:
```

File: lib/cloud.proxy.js [hunk 33]
```diff
@@ -53,13 +53,72 @@ function IoTKitCloud(logger, deviceId, customProxy) {
     me.logger.info("Cloud Proxy created with Cloud Handler: " + me.proxy.type);
     //set mqtt proxy if PROVIDED
     if (conf.connector.mqtt != undefined) {
-        me.mqttProxy = require('@open-iot-service-platform/oisp-sdk-js')(conf).lib.proxies.getMQTTConnector();
-        me.logger.info("Configuring MQTT...");
-        if (deviceConf.device_id && deviceConf.device_token) {
-            me.mqttProxy.updateDeviceInfo(deviceConf);
+    /** 
+    * Checking if SparkplugB is also enabled or not
+    */
```
Comment (srware) [inline lib/cloud.proxy.js:58]: Indentation is incorrect for comment.

**Review by srware — CHANGES_REQUESTED** [review]

File: README.md (hunk 32 above)
Comment (yshashix) [reply README.md:116]: I agree with this comment

File: README.md [hunk 34]
```diff
@@ -109,6 +109,62 @@ To start the oisp-agent service simply execute the start script:
...
+                value: 123
+            }],
+            seq: 1
+         };
+```
+
+3. Sending NGSI-LD compatible data over Mqtt-SpB
+- Register devices for custom Relationship and Property catalog
+- Message must be sent over SpB-Mqtt, which means flag in config must be like "sparkplugB": true,
+
```
[the remark this answers is not in the document]
Comment (yshashix) [reply README.md:155]: yes can be ommited with explicitly stating in heading that it needs SparkplugB enabled

File: lib/cloud.proxy.js (hunk 33 above)
Comment (yshashix) [reply lib/cloud.proxy.js]: changed

**Review by scooter1556 — APPROVED** [review]

**Review by srware — APPROVED** [review]

**Pull request merged**